### PR TITLE
Implement new event hierarchy handling with ClassWalker

### PR DESCRIPTION
**Pull request by MinnDevelopment**
[contributing]: https://github.com/DV8FromTheWorld/JDA/wiki/5%29-Contributing

## Pull Request Etiquette

- [x] I have checked the PRs for upcoming features/bug fixes.
- [x] I have read the [contributing guidelines][contributing].

### Changes

- [x] Internal code
- [ ] Library interface (affecting end-user code) 
- [ ] Documentation
- [ ] Other: \_____ 

Closes Issue: NaN

## Description

This implements simpler handling of event hierarchy in ListenerAdapter and AnnotatedEventManager. This reduces complexity and simplifies the logic for adding new events in the future.

The ClassWalker is a hierarchy traversal algorithm that uses [level order](https://en.wikipedia.org/wiki/Tree_traversal#Breadth-first_search_/_level_order) to iterate both interfaces and classes.

### CODE REVIEW

File: src/main/java/net/dv8tion/jda/api/hooks/ListenerAdapter.java [hunk 1]
```diff
@@ -151,15 +151,27 @@ public void onGuildMemberLeave(@Nonnull GuildMemberLeaveEvent event) {}
     @DeprecatedSince("4.2.0")
     public void onSelfUpdateEmail(@Nonnull SelfUpdateEmailEvent event) {}
 
+    @Deprecated
+    @ForRemoval
+    @DeprecatedSince("4.3.0")
```
Comment (MinnDevelopment) [inline src/main/java/net/dv8tion/jda/api/hooks/ListenerAdapter.java:156]:
```suggestion
    @DeprecatedSince("4.2.1")
```

File: src/main/java/net/dv8tion/jda/api/hooks/ListenerAdapter.java [hunk 2]
```diff
@@ -151,15 +151,27 @@ public void onGuildMemberLeave(@Nonnull GuildMemberLeaveEvent event) {}
...
+    @ReplaceWith("onResumed(ResumedEvent)")
+    public void onResume(@Nonnull ResumedEvent event) {}
+
+    @Deprecated
+    @ForRemoval
+    @DeprecatedSince("4.3.0")
```
Comment (MinnDevelopment) [inline src/main/java/net/dv8tion/jda/api/hooks/ListenerAdapter.java:162]:
```suggestion
    @DeprecatedSince("4.2.1")
```

File: src/main/java/net/dv8tion/jda/api/hooks/ListenerAdapter.java [hunk 3]
```diff
@@ -375,6 +387,13 @@ public final void onEvent(@Nonnull GenericEvent event)
         onGenericEvent(event);
         if (event instanceof UpdateEvent)
             onGenericUpdate((UpdateEvent<?, ?>) event);
+
+        //TODO: Remove once deprecated methods are removed
+        if (event instanceof ResumedEvent)
+            onResume((ResumedEvent) event);
+        else if (event instanceof ReconnectedEvent)
+            onReconnected((ReconnectedEvent) event);
```
Comment (MinnDevelopment) [inline src/main/java/net/dv8tion/jda/api/hooks/ListenerAdapter.java:395]:
```suggestion
            onReconnect((ReconnectedEvent) event);
```

**Review by DV8FromTheWorld — APPROVED** [review]

**Comment by DV8FromTheWorld** [comment]
Event hierarchy execution order between this implementation and the previous ListenerAdapter implementation is identical and was validated via this code test: https://gist.github.com/DV8FromTheWorld/531712ade36fb0525938c3522fdaeb51

**Pull request merged**